frame/whitelist/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+
+		/// Required origin for dispatching whitelisted call with root origin.
+		type DispatchWhitelistedOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The handler of pre-images.
+		// NOTE: recipient is only needed for benchmarks.
+		type PreimageProvider: PreimageProvider<Self::Hash> + PreimageRecipient<Self::Hash>;
+
+		/// The weight information for this pallet.
+		type WeightInfo: weights::WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		CallWhitelisted { call_hash: T::Hash },
+		WhitelistedCallRemoved { call_hash: T::Hash },
+		WhitelistedCallDispatched { call_hash: T::Hash, result: DispatchResultWithPostInfo },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The preimage of the call hash could not be loaded.
+		UnavailablePreImage,
+		/// The call could not be decoded.
+		UndecodableCall,
+		/// The weight of the decoded call was higher than the witness.
+		InvalidCallWeightWitness,
+		/// The call was not whitelisted.
+		CallIsNotWhitelisted,
+		/// The call was already whitelisted; No-Op.
+		CallAlreadyWhitelisted,
+	}
+
+	#[pallet::storage]
+	pub type WhitelistedCall<T: Config> = StorageMap<_, Twox64Concat, T::Hash, (), OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::whitelist_call())]
+		pub fn whitelist_call(origin: OriginFor<T>, call_hash: T::Hash) -> DispatchResult {
+			T::WhitelistOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				!WhitelistedCall::<T>::contains_key(call_hash),
+				Error::<T>::CallAlreadyWhitelisted,
+			);
+
+			WhitelistedCall::<T>::insert(call_hash, ());
+			T::PreimageProvider::request_preimage(&call_hash);
+
+			Self::deposit_event(Event::<T>::CallWhitelisted { call_hash });
+
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::remove_whitelisted_call())]
+		pub fn remove_whitelisted_call(origin: OriginFor<T>, call_hash: T::Hash) -> DispatchResult {
+			T::WhitelistOrigin::ensure_origin(origin)?;
+
+			WhitelistedCall::<T>::take(call_hash).ok_or(Error::<T>::CallIsNotWhitelisted)?;
+
+			T::PreimageProvider::unrequest_preimage(&call_hash);
+
+			Self::deposit_event(Event::<T>::WhitelistedCallRemoved { call_hash });
+
+			Ok(())
+		}
+
+		#[pallet::weight(
+			T::WeightInfo::dispatch_whitelisted_call().saturating_add(*call_weight_witness)
+		)]
+		pub fn dispatch_whitelisted_call(
+			origin: OriginFor<T>,
+			call_hash: T::Hash,
+			call_weight_witness: Weight,
+		) -> DispatchResultWithPostInfo {
+			T::DispatchWhitelistedOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				WhitelistedCall::<T>::contains_key(call_hash),
+				Error::<T>::CallIsNotWhitelisted,
+			);
+
+			let call = T::PreimageProvider::get_preimage(&call_hash)
+				.ok_or(Error::<T>::UnavailablePreImage)?;
+
+			let call = <T as Config>::Call::decode_all_with_depth_limit(
+				sp_api::MAX_EXTRINSIC_DEPTH,
```

Comment:
For completeness we may need to add a bench which tests it with nested calls.  
Currently we only bench the length of an encoded call, but if nesting is more expensive than length, it would be missed.

Reply:
I dont think nesting makes a difference here. The depth limit here is to prevent against stackoverflow and other kinds of call decoding bombs. But the actual computational overhead here should be the same with anything being encoded or decoded really.

And then the weight of the call can help us determine cheaply and ahead of time that the call itself is okay.